File: benchmarks/solutions/zig/a3-valid-parens.zig
const std = @import("std");

fn isValid(s: []const u8) bool {
    var stack: [10000]u8 = undefined;
    var top: usize = 0;

    for (s) |c| {
        switch (c) {
            '(', '[', '{' => {
                stack[top] = c;
                top += 1;
            },
            ')', ']', '}' => {
                if (top == 0) return false;
                top -= 1;
                const expected: u8 = switch (c) {
                    ')' => '(',
                    ']' => '[',
                    '}' => '{',
                    else => unreachable,
                };
                if (stack[top] != expected) return false;
            },
            else => {},
        }
    }
    return top == 0;
}

File: benchmarks/solutions/cpp/a3-valid-parens.cpp
#include <stack>
#include <string>

bool is_valid(const std::string& s) {
    std::stack<char> stk;
    for (char c : s) {
        if (c == '(' || c == '[' || c == '{') {
            stk.push(c);
        } else {
            if (stk.empty()) return false;
            char expected = c == ')' ? '(' : c == ']' ? '[' : '{';
            if (stk.top() != expected) return false;
            stk.pop();
        }
    }
    return stk.empty();
}
